runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java

```diff
@@ -3081,6 +3089,222 @@ public void testActiveWorkRefresh() throws Exception {
+      boolean isActiveWorkRefresh = true;
+      for (ComputationGetDataRequest computationRequest : request.getRequestsList()) {
+        if (!computationRequest.getComputationId().equals(DEFAULT_COMPUTATION_ID)) {
+          isActiveWorkRefresh = false;
+          continue;
+        }
+        for (KeyedGetDataRequest keyedRequest : computationRequest.getRequestsList()) {
+          if (keyedRequest.getWorkToken() == 0
+              || keyedRequest.getShardingKey() != DEFAULT_SHARDING_KEY
+              || keyedRequest.getValuesToFetchCount() != 0
+              || keyedRequest.getBagsToFetchCount() != 0
+              || keyedRequest.getTagValuePrefixesToFetchCount() != 0
+              || keyedRequest.getWatermarkHoldsToFetchCount() != 0) {
+            isActiveWorkRefresh = false;
+            continue;
+          }
+          for (LatencyAttribution la : keyedRequest.getLatencyAttributionList()) {
+            EnumMap<LatencyAttribution.State, Duration> durations =
+                totalDurations.computeIfAbsent(
+                    keyedRequest.getWorkToken(),
+                    (Long workToken) ->
+                        new EnumMap<LatencyAttribution.State, Duration>(
+                            LatencyAttribution.State.class));
+            Duration old = durations.get(la.getState());
+            Duration cur = Duration.millis(la.getTotalDurationMillis());
+            if (old == null || old.isShorterThan(cur)) {
+              durations.put(la.getState(), cur);
+            }
+          }
+        }
+      }
+      if (!isActiveWorkRefresh) {
+        // The unit test below for state QUEUED relies on this delay.
+        Uninterruptibles.sleepUninterruptibly(2000, TimeUnit.MILLISECONDS);
+      }
+      return EMPTY_DATA_RESPONDER.apply(request);
+    }
+  }
+
+  @Test
+  public void testLatencyAttributionToQueuedState() throws Exception {
+    final int workToken = 323232; // A unique id makes it easier to find logs.
+
+    List<ParallelInstruction> instructions =
+        Arrays.asList(
+            makeSourceInstruction(StringUtf8Coder.of()),
+            makeDoFnInstruction(new SlowDoFn(Duration.millis(2000)), 0, StringUtf8Coder.of()),
+            makeSinkInstruction(StringUtf8Coder.of(), 0));
+
+    FakeWindmillServer server = new FakeWindmillServer(errorCollector);
+    server.setGetDataSleep(Duration.ZERO);
+    StreamingDataflowWorkerOptions options = createTestingPipelineOptions(server);
+    options.setActiveWorkRefreshPeriodMillis(100);
+    // A single-threaded worker processes work sequentially, leaving a second work item in state
+    // QUEUED until the first work item is committed.
+    options.setNumberOfWorkerHarnessThreads(1);
+    StreamingDataflowWorker worker = makeWorker(instructions, options, false /* publishCounters */);
+    worker.start();
+
+    ActiveWorkRefreshSink awrSink = new ActiveWorkRefreshSink();
+    for (int i = 0; i < 1000; ++i) {
+      server.addDataFnToOffer(
```

Comment:
can you modify the server to allow specifying a default? instead of just loading 1000 up

Reply:
Done (after next force push). I've put the default behind a new nested class 'ResponseQueue'. This allows us to reuse this class for GetWork, GetData, and CommitWork requests in unit test. I hope that's not overengineered.